### DISCUSSION TITLE
Set GPU P-state within settings.

### DISCUSSION THREAD

**Discussion by vidiotgameboss** [vladmandic/sdnext#1254] [Ideas]
### Feature description

Add options within automatic to control the GPU so that only during generation it increases performance and during idle it cools off more.

This would not be used by many casual users but I believe overclockers/powerusers or just people who care about their PC's thermals would benefit immensely in the way that they would run out their GPU's lifespan less sooner and pay less on electricity bills. The reasoning for it is that while a casual user would already do this by having their GPU dynamically change power states, power users normally lock their GPU clocks very tightly and would force P8/P12 on desktop, with only certain apps allowing higher power states, normally this can be defined within software by having a list of running processes which can set P0, otherwise the GPU goes into its lowest P state. 

In SD webui's case, this would mean adding python.exe to the list of software that sets the GPU to P0 which in turn would immediately lock the GPU at P0 and force it to work full time which is not needed until generation happens. If SD could instead change the GPU Pstate on its own at generation that would be good because after generation the GPU would have time to cool off more in P8/P12.

I write this as more of a niche request for all powerusers, overclockers, thermal OCD people. I and some others I know personally do the above to our GPUs on top of undervolting our GPUs and limiting power so that when it goes P0 it can be more efficient.

**Comment by vladmandic** [comment]
no chance i'd put something like that in the app. but its totally valid for user to have it in custom launcher.
for example:

> webui-user.sh
```batch
sudo nvidia-smi -pm ENABLED -i 0
sudo nvidia-smi -ac 3004,875 -i 0
./webui.sh 
sudo nvidia-smi -rac -i 0
```

which
- set persistent mode
- set gpu clock to specific value
- run webui
- reset gpu to defaults on exit

btw, i'm going to convert this to discussion as it may be interesting for others, but its not going to be done in the app itself.

**Comment by cool-bigdogs-tshirt** [comment]
huh... this conversation has led me to discover that my GPU has been stuck in P2, which is apparently forced for CUDA and used to only be toggle-able in windows?

gonna copy/paste vlad's commands verbatim and watch for flames

**Reply by vladmandic** [reply]
Most likely it's not going to work - you need to dump available clocks first and then select one that you want from the list, you can just a random value (values I've put in are from A100).

**Reply by cool-bigdogs-tshirt** [reply]
yeah, well, i'm on wayland anyways, so... thanks nvidia... you can only set 'coolbits' in X config
`Setting applications clocks is not supported for GPU`

surprised those clocks are so low though... i guess the benefit is in the vram? let me know if you have a spare ;)

**Reply by vidiotgameboss** [reply]
on windows you can set those in a variety of ways unlike linux. everytime ive went on linux ive been met with a lack of software, no msi afterburner, no rivatuner rtss, no nvidia inspector with its nvidia multi monitor to control its p states or things like limit-nvpstate and no equivalent to software that i use often for certain things, so i stick with windows despite linux doing many things better than windows

if youre on linux youll have less control over your gpu and cpu imo, i think amd gpus may offer more control but i dont know for sure

**Reply by vladmandic** [reply]
> surprised those clocks are so low though... i guess the benefit is in the vram? let me know if you have a spare ;)

Yeah, A100 is a workhorse, not a racehorse - slow and massively scalable. For data centers, I'd say my sweet spot right now would be L40 if you can find it anywhere.
